ArjunaCore/arjuna/classes/com/arjuna/ats/arjuna/recovery/ExtendedRecoveryModule.java

```diff
@@ -0,0 +1,50 @@
+/*
+ * JBoss, Home of Professional Open Source.
+ * Copyright 2018, Red Hat, Inc., and individual contributors
+ * as indicated by the @author tags. See the copyright.txt file in the
+ * distribution for a full listing of individual contributors.
+ *
+ * This is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation; either version 2.1 of
+ * the License, or (at your option) any later version.
+ *
+ * This software is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this software; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+package com.arjuna.ats.arjuna.recovery;
+
+import java.util.Set;
+
+public interface ExtendedRecoveryModule extends RecoveryModule {
+    /**
+     * Ask the module for the names of any resources it was unable to make
+     * contact with during the last recovery pass.
+     *
+     * @param uncontactedResources a set that will contain the resource names of
+     *                             any resources that the module could not contact
+     * @return the number of resources that were added to the input set
+     */
+    default int getUncontactedResourceNames(Set<String> uncontactedResources) {
```

Comment:
No, the intended behaviour is to iterate over all modules gathering up the results as we go which is more efficient that creating a new map per recovery module and then copying over the data into a holding map.

Reply:
I see. Just I'm not sure if i put it understandable. My point was about way how the api is designed - by not using parameter to mutate but returning the list of the resources.

Reply:
Two points:
- in view of the comments so far I plan to remove the interface in favour of something simpler
- the interface was designed for performance since the method was to be called during recovery which iterates over each recovery module gathering up resources. Using your technique would involve lots of map creations and copying between maps. The proposed interface allows periodic recover to create one map and have all the results in a singe place (thus reducing memory, CPU cycles and code complexity)